pkg/cache/cache.go

```diff
@@ -721,6 +725,18 @@ func (c *Cache) MatchingClusterQueues(nsLabels map[string]string) sets.Set[strin
 	return cqs
 }
 
+// ShouldCheckWorkloadsPreemprion returns true if it has workloads pending preemption now.
+func (c *Cache) ShouldCheckWorkloadsPreemprion() bool {
+	c.RLock()
+	defer c.RUnlock()
+	for _, cq := range c.clusterQueues {
```

Comment:
The problem with this is that it will get triggered a lot more (at least for workload and admission check updates), an depending on the current queue state and the change taking place we might need to do the same iteration.

On the plus side for the current implementation, the number of calls is limited to the number of preemption runs, on average we should not have more that 2/sec, while only holding the read lock and it will return on the first queue that needs preemption.

Reply:
> The problem with this is that it will get triggered a lot more (at least for workload and admission check updates), an depending on the current queue state and the change taking place we might need to do the same iteration.

 We already keep track and take lock for workload updates in cache in `cache.UpdateWorkload`. This is the strategy we use to keep the `podsReady` condition in-sync. 

> On the plus side for the current implementation, the number of calls is limited to the number of preemption runs, on average we should not have more that 2/sec, while only holding the read lock and it will return on the first queue that needs preemption.

I see, but I'm concern with the worst case scenario of iteration over all workloads. There might be users where it counts in millions. FYI @alculquicondor, @mwielgus ?

Reply:
Hundred thousand is also a bit of a stretch at this moment. Realistically, I would rather expect  small two digit thousand workloads being queued at  the same time.  

Obviously the loop over all workloads could be improved and with the proper workload status update handling - completely avoided. The question is - does it matter enough at this moment. With the 2 stage admission being quite complex, i would rather suggest documenting all non-critical bottlenecks with:

* how bad they are
* how often they are triggered and how can we limit them
* what happens if they are triggered
* how can they be fixed

instead of trying to fix them all, in one uber-complex PR.

Reply:
I'm fine with just documenting (probably a good comment in code would suffice) the potential bottlenecks. Especially, as the code makes the fact we iterate over all workloads hidden, as the loops are on different levels.

Reply:
done

Reply:
Please document in an issue, and reference back here.